### PR TITLE
[AArch64] When hardening against SLS, only create called thunks

### PR DESCRIPTION
**Pull request by atrosinenko**
In preparation for implementing hardening of BLRA* instructions, restrict thunk function generation to only the thunks being actually called from any function. As described in the existing comments, emitting all possible thunks for BLRAA and BLRAB instructions would mean adding about 1800 functions in total, most of which are likely not to be called.

This commit merges AArch64SLSHardening class into SLSBLRThunkInserter, so thunks can be created as needed while rewriting a machine function. The usages of TII, TRI and ST fields of AArch64SLSHardening class are replaced with requesting them in-place, as ThunkInserter assumes multiple "entry points" in contrast to the only runOnMachineFunction method of AArch64SLSHardening.

The runOnMachineFunction method essentially replaces pre-existing insertThunks implementation as there is no more need to insert all possible thunks unconditionally. Instead, thunks are created on first use from inside of insertThunks method.

### CODE REVIEW

File: llvm/test/CodeGen/AArch64/speculation-hardening-sls-blr.mir
Comment (kbeyls) [inline llvm/test/CodeGen/AArch64/speculation-hardening-sls-blr.mir:56]: I think it is important to check that the body of the thunk inserted is as expected.
What is the motivation to remove that check, and only check that a thunk with the expected name was inserted, without checking its body?

File: llvm/lib/Target/AArch64/AArch64SLSHardening.cpp
Comment (kbeyls) [inline llvm/lib/Target/AArch64/AArch64SLSHardening.cpp:43]: I'm not sure if `SLSBLRThunkInserter` is the most appropriate name for this class.
IIUC, this class will also harden Returns and BRs, (see method `hardenReturnsAndBRs`), which does not insert thunks.
If I understand this correctly, I think sticking with the original name of "AArch64SLSHardening" describes better what this class does?

File: llvm/lib/Target/AArch64/AArch64SLSHardening.cpp
Comment (kbeyls) [inline llvm/lib/Target/AArch64/AArch64SLSHardening.cpp:41]: I guess once support for BLRA* instruction will be added, the bitset may contain up to 29\*29-ish registers and this typedef will change to something different than a `uint32_t`?
I agree this can be done in the follow-on patch that will support BLRA* instructions.

File: llvm/test/CodeGen/AArch64/speculation-hardening-sls-blr.mir
Comment (atrosinenko) [reply llvm/test/CodeGen/AArch64/speculation-hardening-sls-blr.mir:56]: The above lines are **input** to `llc`, so I removed them to not conflict with the thunks that are inserted by the pass. Unfortunately, MIR syntax is quite verbose, so I added a few basic CHECK lines testing MIR output and duplicated the RUN lines to generate and check assembly output as well.

Now there is no separation between function-rewriting and thunk-inserting passes, and I assumed it is safe to expect that initially there are no thunks in the module. I guess the original version of thunk insertion logic would assert on pre-existing thunks as well (due to unknown thunk names with `.1` suffix).

File: llvm/lib/Target/AArch64/AArch64SLSHardening.cpp
Comment (atrosinenko) [reply llvm/lib/Target/AArch64/AArch64SLSHardening.cpp:41]: Yes, `ThunkSet` will be class.

File: llvm/lib/Target/AArch64/AArch64SLSHardening.cpp
Comment (atrosinenko) [reply llvm/lib/Target/AArch64/AArch64SLSHardening.cpp]: Renamed it to `SLSHardeningInserter`, so the `AArch64SLSHardening` name is kept for the pass itself.

**Review by atrosinenko — COMMENTED** [review]
@kbeyls thank you for the comments!

File: llvm/lib/Target/AArch64/AArch64SLSHardening.cpp
Comment (atrosinenko) [reply llvm/lib/Target/AArch64/AArch64SLSHardening.cpp:41]: Here is the PR: #97605

File: llvm/test/CodeGen/AArch64/speculation-hardening-sls-blr.mir
Comment (kbeyls) [reply llvm/test/CodeGen/AArch64/speculation-hardening-sls-blr.mir:56]: ah, yes, I was confused and didn't spot these were input lines....
I looked a bit more now at all of the existing SLS-hardening tests, and the content of the thunk is already tested in https://github.com/llvm/llvm-project/blob/f15266e9d5d87a291a515064f8f0dd5f0c412ed3/llvm/test/CodeGen/AArch64/speculation-hardening-sls.ll#L195

So, there isn't really a need to test for it separately here.
Apologies for giving you more work here!

I think it's fine for the test to go in as is; and it's equally fine for the test to go in without the CHECK lines for the content of the thunk. I'll leave it to you to make that decision.

**Review by kbeyls — APPROVED** [review]
LGTM, thanks! (after fixing the clang-format failure)

**Pull request merged**
